File: server/Center/Node/NodeManager.cs
//////////////////////////////////////////////////////////////////////////
// 
// 文件：Center/Node/NodeManager.cs
// 作者：Xoen Xie
// 时间：2023/06/19
// 描述：节点管理
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#pragma warning disable CS8603, CS8600, CS8602, CS8604, CS8618
using XSF;

public class NodeManager : DicNPManager
{
    enum RunStep
    {
        None = 0,
        StartServer,
        WaitHandshake,
        HandshakeDone,
        OK,
    }

    private List<ServerInfo>[] m_LostList;
    private Dictionary<uint, ServerInfo> m_Nodes;
    private byte[] m_NodeIndex;
    private uint m_nInnerPort;
    private uint m_nOutPort;

    private int m_nStartIndex;
    private RunStep m_nStep;

    private ServerNode m_CurStartNode;

    public NodeManager()
    {   
        m_Nodes = new Dictionary<uint, ServerInfo>();
        m_LostList = new List<ServerInfo>[(int)EP.Max];
        m_NodeIndex = new byte[(int)EP.Max];

        for(int i = 0; i < (int)EP.Max; i ++)
        {
            m_LostList[i] = new List<ServerInfo>();
            m_NodeIndex[i] = 1;
        }
    }

    public override bool Init(ModuleInit init)
    {
        m_nInnerPort = XSFCore.Config.InnerPortStart;
        m_nOutPort = XSFCore.Config.OutPortStart;

        return base.Init(init);
    }

    public override bool Start()
    {
        m_nStartIndex = 0;
        m_nStep = RunStep.StartServer;
        
        return base.Start();
    }

    public override void OnStartClose()
    {
        var message = XSFCore.GetMessage((ushort)XsfPbid.SMSGID.CCcStop);
        Broadcast(message, 0);
    }

    public override void DoRegist()
    {
        XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CcCHandshake, new Executor_Cc_C_Handshake());
        XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CcCHeartbeat, new Executor_Cc_C_Heartbeat());
        XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CcCServerOk, new Executor_Cc_C_ServerOk());
    }

    public override ModuleRunCode OnStartCheck() 
    {
        if(!XSFCore.Config.AutoStart)
        {
            return ModuleRunCode.OK;
        }
        else
        {
            switch(m_nStep)
            {
            case RunStep.StartServer:
                {
                    m_CurStartNode = XSFCore.Config.NodeList[m_nStartIndex];

                    Serilog.Log.Information("Start server type={0}",  XSFCore.EP2CNName((byte)m_CurStartNode.ep));

                    string args = $"./single_start.sh {XSFCore.Server.InitData.ServerTag} {XSFCore.EP2Name((byte)m_CurStartNode.ep)} {XSFCore.Server.SID.ID}";
                    XSFCore.StartProcess("sh", args, XSFCore.Server.InitData.WorkDir);

                    Serilog.Log.Information("RunStep.StartServer done ....");

                    m_nStep = RunStep.WaitHandshake;
                }
                break;

            case RunStep.HandshakeDone:
                {
                    m_nStartIndex ++;
                    if(m_nStartIndex >= XSFCore.Config.NodeList.Length)
                    {
                        m_nStep = RunStep.OK;
                    }
                    else
                    {
                        m_nStep = RunStep.StartServer;
                    }
                }
                break;


            case RunStep.OK:
                return ModuleRunCode.OK; 
            } 

            return ModuleRunCode.Wait;
        }
    }

    public ServerInfo AddNode(uint nID, string ip, uint[] ports)
    {
        ServerID sid = ServerID.GetSID(nID);
        ServerInfo newNode = null;
        if(sid.Index == 0 )
        {
            var nodeLost = GetLostNode(sid.Type, nID, false);
            if(nodeLost == null)
            {
                newNode = new ServerInfo();
                GetPort(sid.Type, ref newNode.Ports);

                sid.Index = GetIndex(sid.Type);
                newNode.ID = ServerID.GetID(sid);
                newNode.IP = ip;
                newNode.Status = NodeStatus.New;

                Serilog.Log.Information($"NodeManager AddNode new, id={newNode.ID}, ip={ip}");
            }
            else
            {
                newNode = nodeLost;
                newNode.Status = NodeStatus.New;
                newNode.IP = ip;

                Serilog.Log.Information($"NodeManager AddNode new, find lost server node id={newNode.ID}, ip1={newNode.IP}, ip2={ip}");
            }

            m_Nodes[newNode.ID] = newNode;
        }
        else
        {
            if(m_Nodes.TryGetValue(nID, out newNode))
            {
                Serilog.Log.Information($"NodeManager AddNode find exist 1, id={nID}, ip={ip}");
            }
            else
            {
                var nodeLost = GetLostNode(sid.Type, nID, true);
                if(nodeLost == null)
                {
                    Serilog.Log.Error($"NodeManager AddNode nodeLost not found, id={nID}, ip={ip}");
                    return null;
                }
                else
                {
                    if(!nodeLost.IP.Equals(ip))
                    {
                        Serilog.Log.Error($"NodeManager AddNode find exist, ip error, id={nID}, ip1={nodeLost.IP}, ip2={ip}");
                        return null;
                    }

                    for(int i = 0; i < (int)EP.Max; i ++)
                    {
                        if(nodeLost.Ports[i] != ports[i])
                        {
                            Serilog.Log.Error($"NodeManager AddNode find exist, port error, id={nID}, i={i} port1={nodeLost.Ports[i]}, port2={ports[i]}");
                            return null;
                        }
                    }
                }

                newNode = nodeLost;
                m_Nodes.Add(nID, newNode);
                Serilog.Log.Information($"NodeManager AddNode find exist 2, id={nID}, ip={ip}");
            }
        }

        return newNode;
    }

    public byte GetIndex(byte nEP)
    {
        var index = m_NodeIndex[nEP];
        m_NodeIndex[nEP] ++;

        return index;
    }

    public void GetPort(byte nEP, ref uint[] ports)
    {
        switch((EP)nEP)
        {
        case EP.Gate:
            ports[(int)EP.Client] = GetNextPort(false);
            break;

        case EP.Game:
            ports[(int)EP.Gate] = GetNextPort(true);
            break;

        case EP.Login:
            ports[(int)EP.Gate] = GetNextPort(true);
            break;

        case EP.DB:
            ports[(int)EP.DB] = GetNextPort(true);
            break;

        case EP.Hub:
            ports[(int)EP.Hub] = GetNextPort(true);
            break;

        default:
            break;
        }
    }

    private uint GetNextPort(bool IsInner)
    {
        if(IsInner)
        {
            return m_nInnerPort ++;
        } 
        else 
        {
            return m_nOutPort ++;
        }
    }

    public ServerInfo GetLostNode(byte nEP, uint nID, bool CheckEqual)
    {
        var list = m_LostList[nEP];
        for(int i = 0; i < list.Count; i ++)
        {
            var node = list[i];
            if(CheckEqual)
            {
                if(node.ID == nID)
                {
                    list.RemoveAt(i);
                    return node;
                }
            }
            else
            {
                list.RemoveAt(i);
                return node;
            }
        }

        return null;
    }

    public override void OnNPConnected(NetPoint np)
    {
        var message = XSFCore.GetMessage((ushort)XsfPbid.SMSGID.CCcServerInfo) as XsfMsg.MSG_C_Cc_ServerInfo;
        message.mPB.Infos.Clear();

        // 把当前已经收到的服务器信息下发给新加入的节点
        uint nID = np.ID;
        //Serilog.Log.Information("OnNPConnected nID=" + nID);
        foreach (KeyValuePair<uint, ServerInfo> kvp in m_Nodes)
        {
            if(kvp.Key != nID)
            {
                var node = kvp.Value;
                XsfPb.MSG_ServerInfo info = new XsfPb.MSG_ServerInfo();
                info.ServerId = node.ID;
                info.Ip = node.IP;
                for(int i = 0; i < node.Ports.Length; i ++)
                    info.Ports.Add(node.Ports[i]);
                info.Status = (uint)node.Status;
                message.mPB.Infos.Add(info);
            }
        }

        if(message.mPB.Infos.Count > 0)
            np.SendMessage(message);

        // 把新加入的节点信息，广播给其他所有服务器节点
        {
            message.mPB.Infos.Clear();

            ServerInfo nodeAdd = null;
            m_Nodes.TryGetValue(np.ID, out nodeAdd);
            XsfPb.MSG_ServerInfo info = new XsfPb.MSG_ServerInfo();
            info.ServerId = nodeAdd.ID;
            info.Ip = nodeAdd.IP;
            for(int i = 0; i < nodeAdd.Ports.Length; i ++)
                info.Ports.Add(nodeAdd.Ports[i]);
            info.Status = (uint)nodeAdd.Status;
            message.mPB.Infos.Add(info);
                    
            Broadcast(message, np.ID);
        }

        if(m_CurStartNode != null && (byte)m_CurStartNode.ep == np.SID.Type)
        {
            m_nStep = RunStep.HandshakeDone;
        }
    }

    public override void OnNPLost(NetPoint np)
    {
        ServerInfo node = null;
        if(m_Nodes.TryGetValue(np.ID, out node))
        {
            m_Nodes.Remove(np.ID);
            m_LostList[np.SID.Type].Add(node);
        }
        else
        {
            Serilog.Log.Information("NodeManager OnNPLost, node not exist, id={0}", np.ID);
            return;
        }

        var message = XSFCore.GetMessage((ushort)XsfPbid.SMSGID.CCcServerLost) as XsfMsg.MSG_C_Cc_ServerLost;
        message.mPB.ServerId = np.ID;
        Serilog.Log.Information("【中心服】有服务器节点离线, id={0}", np.ID);

        Broadcast(message, 0);
    }

    public void OnNodeOk(uint nID)
    {
        ServerInfo node = null;
        if(m_Nodes.TryGetValue(nID, out node))
        {
            node.Status = NodeStatus.Ok;
        }
        else
        {
            Serilog.Log.Information("NodeManager OnNodeOk, node not exist, id={0}", nID);
            return;
        }

        Serilog.Log.Information("【中心服】收到服务器已准备好 id={0}", nID);

        var message = XSFCore.GetMessage((ushort)XsfPbid.SMSGID.CCcServerOk) as XsfMsg.MSG_C_Cc_ServerOk;
        message.mPB.ServerId = nID;

        Broadcast(message, nID);
    }




    public static void CreateModule()
    {
        NodeManager module = new NodeManager();

        NPManagerInit init = new NPManagerInit();
        init.ID = (int)ModuleID.Node;
        init.Name = "NodeManager";
        init.NoWaitStart = true;
        init.Port = (int)XSFCore.Config.CenterPort;

        XSFCore.Server.AddModule(module, init);
    }

    public static NodeManager Instance
    {
        get
        {
            return XSFCore.Server.GetModule((int)ModuleID.Node) as NodeManager;
        }
    }
}

File: server-cpp/source/center/server/src/NodeManager.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：source/center/server/src/NodeManager.cpp
// 作者：Xoen Xie
// 时间：2020/02/29
// 描述：节点管理
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#include "NodeManager.h"
#include "XSF.h"
#include "DMessageInc.h"
#include "CenterExecutors.h"
#include "NetPoint.h"
#include "Server.h"

bool NodeManager::Init(ModuleInit * pInit) 
{
    auto pConfig = XSFCore::GetServer()->GetConfig();
    m_nInnerPort = pConfig->InnerPortStart;
    m_nOutPort = pConfig->OutPortStart;

    for(uint8 i = 0; i < EP_Max; i ++)
    {
        m_NodeIndex[i] = 1;
    }

    return NormalNetPointManager::Init(pInit);
}

bool NodeManager::Start(void) 
{
    m_nStartIndex = 0;
    m_nStep = RunStep_StartServer;

    return NormalNetPointManager::Start();
}

void NodeManager::OnClose(void)
{
    auto pMessage = XSFCore::GetMessage(xsf_pbid::SMSGID::C_Cc_Stop);
    Broadcast(pMessage, 0);
}

void NodeManager::DoRegist(void)
{
    XSFCore::SetMessageExecutor(xsf_pbid::SMSGID::Cc_C_Handshake, new Executor_Cc_C_Handshake());
    XSFCore::SetMessageExecutor(xsf_pbid::SMSGID::Cc_C_Heartbeat, new Executor_Cc_C_Heartbeat());
    XSFCore::SetMessageExecutor(xsf_pbid::SMSGID::Cc_C_ServerOk, new Executor_Cc_C_ServerOk());
}

uint8 NodeManager::OnStartCheck()
{
    auto pConfig = XSFCore::GetServer()->GetConfig();
    auto pInitData = XSFCore::GetServer()->GetInitData();
    if(!pConfig->AutoStart)
    {
        XSF_INFO("NodeManager::OnStartCheck no auto start ...");
        return ModuleRunCode_OK;
    }
    else
    {
        switch (m_nStep)
        {
        case RunStep_StartServer:
            {
                const ServerNode & node = pConfig->NodeList[m_nStartIndex];
                m_nCurStartEP = node.nEP;

                char sShell[BUFFER_SIZE_1024] = {0};
                sprintf(sShell, "cd ../; sh ./single_start.sh %s %s %u", pInitData->sTag, xsf::EP2Name(m_nCurStartEP), XSFCore::GetServer()->GetSID()->S.server);
                XSFCore::ShellExec(sShell);
                m_nStep = RunStep_WaitHandshake;
            }
            break;

        case RunStep_HandshakeDone:
            {
                m_nStartIndex ++;
                if(m_nStartIndex >= pConfig->NodeList.size())
                {
                    m_nStep = RunStep_OK;
                }
                else
                {
                    m_nStep = RunStep_StartServer;
                }
            }
            break;
        
        case RunStep_OK:
            return ModuleRunCode_OK;
        }
    }

    return ModuleRunCode_Wait;
}


void NodeManager::OnNetPointLost( NetPoint * pPoint )
{
    ServerInfo * pInfo = nullptr;
    auto it = m_Nodes.find(pPoint->GetSID()->ID);
    if(it != m_Nodes.end())
    {
        pInfo = it->second;
        m_Nodes.erase(it);
        m_LostList[pPoint->GetSID()->S.type].push_back(pInfo);
    }
    else
    {
        XSF_INFO("NodeManager::OnNetPointLost node not exist, id=%u", pPoint->GetSID()->ID);
        return;
    }

    XSF_CAST(pMessage, XSFCore::GetMessage(xsf_pbid::SMSGID::C_Cc_ServerLost), xsf_msg::MSG_C_Cc_ServerLost);
    pMessage->mPB.set_server_id(pPoint->GetSID()->ID);
    XSF_INFO("【中心服】有服务器节点离线, id=%u", pPoint->GetSID()->ID);
    Broadcast(pMessage, 0);
}

void NodeManager::OnNetPointConnected( NetPoint * pPoint )
{
    XSF_CAST(pMessage, XSFCore::GetMessage(xsf_pbid::SMSGID::C_Cc_ServerInfo), xsf_msg::MSG_C_Cc_ServerInfo);
    pMessage->mPB.clear_infos();
    uint32 nID = pPoint->GetSID()->ID;

    // 把当前已经收到的服务器信息下发给新加入的节点
    for(auto it = m_Nodes.begin(); it != m_Nodes.end(); it ++)
    {
        if(it->first != nID)
        {
            auto pInfo = pMessage->mPB.add_infos();
            pInfo->set_server_id(it->second->ID);
            pInfo->set_ip(it->second->sIP);
            pInfo->set_status(it->second->Status);

            for(uint8 i = 0; i < EP_Max; i ++)
                pInfo->add_ports(it->second->Ports[i]);
        }
    }

    if(pMessage->mPB.infos_size() > 0)
    {
        pPoint->SendMessage(pMessage);
    }

    // 把新加入的节点信息，广播给其他所有服务器节点
    pMessage->mPB.clear_infos();
    auto pNode = m_Nodes.find(nID)->second;
    auto pInfo = pMessage->mPB.add_infos();
    pInfo->set_server_id(pNode->ID);
    pInfo->set_ip(pNode->sIP);
    pInfo->set_status(pNode->Status);

    for(uint8 i = 0; i < EP_Max; i ++)
        pInfo->add_ports(pNode->Ports[i]);

    Broadcast(pMessage, nID);

    if(m_nCurStartEP == pPoint->GetSID()->S.type)
    {
        m_nStep = RunStep_HandshakeDone;
    }
}



ServerInfo * NodeManager::AddNode(uint32 nID, const char * ip, vector<uint> & ports)
{
    ServerInfo * pNewNode = nullptr;
    SID sid;
    sid.ID = nID;
    XSF_INFO("NodeManager::AddNode id=%u %u-%s-%u", nID, sid.S.server, EP2CNName(sid.S.type), sid.S.index);

    if(sid.S.index == 0)
    {
        ServerInfo * nodeLost = GetLostNode(sid.S.type, nID, false);
        if(nodeLost == nullptr)
        {
            pNewNode = new ServerInfo();
            GetPort(sid.S.type, pNewNode);
            sid.S.index = GetIndex(sid.S.type);

            pNewNode->ID = sid.ID;
            strcpy(pNewNode->sIP, ip);
            pNewNode->Status = NodeStatus_New;

            XSF_INFO("NodeManager::AddNode Add new id=%u, ip=%s", pNewNode->ID, ip); 
        }
        else
        {
            pNewNode = nodeLost;
            pNewNode->Status = NodeStatus_New;
            strcpy(pNewNode->sIP, ip);
            XSF_INFO("NodeManager::AddNode Add new, find lost server node, id=%u, ip lost=%s, ip=%s", pNewNode->ID, pNewNode->sIP, ip); 
        }

        if(!m_Nodes.insert(InfoMap::value_type(pNewNode->ID, pNewNode)).second)
        {
            XSF_ERROR("NodeManager::AddNode m_Nodes.insert insert error");
            return nullptr;
        }
    }
    else
    {
        auto it = m_Nodes.find(nID);
        if(it != m_Nodes.end())
        {
            pNewNode = it->second;
            XSF_INFO("NodeManager AddNode find exist 1, id=%u, ip=%s", nID, ip);
        }
        else
        {
            ServerInfo * nodeLost = GetLostNode(sid.S.type, nID, true);
            if(nodeLost == nullptr)
            {
                XSF_ERROR("NodeManager AddNode nodeLost not found, id=%u, ip=%s", nID, ip);
                return nullptr;
            }
            else
            {
                if (strncmp(pNewNode->sIP, ip, strlen(ip)) != 0)
                {
                    XSF_ERROR("NodeManager AddNode ip error, id=%u, node ip=%s, ip=%s", nID, pNewNode->sIP, ip);
                    return nullptr;
                }

                for(uint32 i = 0; i < EP_Max; i ++)
                {
                    if(pNewNode->Ports[i] != ports[i])
                    {
                        XSF_ERROR("NodeManager AddNode port error, id=%u, i=%u, node port=%u, port=%u", nID, i, pNewNode->Ports[i], ports[i]);
                        return nullptr;
                    }
                }
            }

            pNewNode = nodeLost;
            if(!m_Nodes.insert(InfoMap::value_type(pNewNode->ID, pNewNode)).second)
            {
                XSF_ERROR("NodeManager::AddNode m_Nodes.insert insert error 2");
                return nullptr;
            }
        }
    }

    return pNewNode;
}

void NodeManager::OnNodeOK(uint32 nID)
{
    auto it = m_Nodes.find(nID);
    if(it != m_Nodes.end())
    {
        it->second->Status = NodeStatus_Ok;
    }
    else
    {
        XSF_WARN("NodeManager::OnNodeOK, node not exist, id=%u", nID);
        return;
    }

    XSF_INFO("【中心服】收到服务器已准备好 id=%u", nID);

    XSF_CAST(pMessage, XSFCore::GetMessage(xsf_pbid::SMSGID::C_Cc_ServerOk), xsf_msg::MSG_C_Cc_ServerOk);
    pMessage->mPB.set_server_id(nID);
    Broadcast(pMessage, nID);
}

ServerInfo * NodeManager::GetLostNode(byte nEP, uint32 nID, bool CheckEqual)
{
    list<ServerInfo*> & l = m_LostList[nEP];
    for(auto it = l.begin(); it != l.end(); it ++)
    {
        auto * pInfo = (*it);
        if(CheckEqual)
        {
            if(pInfo->ID == nID)
            {
                l.erase(it);
                return pInfo;
            }
        }
        else
        {
            l.erase(it);
            return pInfo;
        }
    }

    return nullptr;
}

void NodeManager::GetPort(byte nEP, ServerInfo * pInfo)
{
    switch(nEP)
    {
    case EP_Gate:
        pInfo->Ports[EP_Client] = GetNextPort(false);
        break;

    case EP_Game:
        pInfo->Ports[EP_Gate] = GetNextPort(true);
        break;

    case EP_Login:
        pInfo->Ports[EP_Gate] = GetNextPort(true);
        break;

    default:
        break;
    }
}

void SetNodeManager(void)
{
    NetPointManagerInit * init = new NetPointManagerInit();
    init->nModuleID = CenterModule_NodeManager;
    init->NoWaitStart = true;
    strcpy(init->sName, "NodeManager");
    init->nPort = XSFCore::GetServer()->GetConfig()->CenterPort;

    XSFCore::GetServer()->AddModule(NodeManager::Instance(), init);
}